{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample QC for GWAS analysis\n",
    "\n",
    "## As-Is Software Disclaimer\n",
    "\n",
    "This notebook is delivered \"As-Is\". Notwithstanding anything to the contrary, DNAnexus will have no warranty, support, liability or other obligations with respect to Materials provided hereunder.\n",
    "\n",
    "[MIT License](https://github.com/dnanexus/UKB_RAP/blob/main/LICENSE) applies to this notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Jupyterlab app details (launch configuration)\n",
    "\n",
    "Recommended configuration\n",
    "- Runtime: ~5 min\n",
    "- Cluster configuration: `Single Node`\n",
    "- Recommended instance: `mem1_ssd1_v2_x36`\n",
    "- Cost: ~£0.3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dependencies\n",
    "\n",
    "|Library |License|\n",
    "|:------------- |:-------------|\n",
    "|[pandas](https://pandas.pydata.org/) |[BSD-3](https://github.com/pandas-dev/pandas/blob/main/LICENSE)|\n",
    "|[numpy](https://numpy.org/) |[BSD-3](https://github.com/numpy/numpy/blob/main/LICENSE.txt)|\n",
    "\n",
    "## Introduction\n",
    "\n",
    "This notebook:\n",
    "- Loads cohorts created in cohort browser\n",
    "- Performs sample QC\n",
    "- Creates a file containing phenotype and covariate information needed for GWAS analysis\n",
    "\n",
    "Data used for this notebook:\n",
    "- Cohorts (`ischemia_cases` `ischemia_controls`) created using Cohort Browser\n",
    "- Dataset to retrieve phenotype data for both cohorts (`ischemia.pheno`)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare your environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Import packages\n",
    "# dxpy allows python to interact with the platform storage\n",
    "import dxpy\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "import subprocess\n",
    "import glob\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "imputation_folder = 'Imputation from genotype (GEL)'\n",
    "imputation_field_id = '21008'\n",
    "output_dir = '/Data/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load dataset description containing phenotypic data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Automatically discover dispensed dataset ID\n",
    "dispensed_dataset = dxpy.find_one_data_object(\n",
    "    typename='Dataset', name='app*.dataset', folder='/', name_mode='glob'\n",
    ")\n",
    "dispensed_dataset_id = dispensed_dataset['id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get project ID\n",
    "project_id = dxpy.find_one_project()['id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dataset = (':').join([project_id, dispensed_dataset_id])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the `-ddd` parameter will extract 3 dictionary files associated with the dataset.\n",
    "\n",
    "The 3 dictionary files that are returned include:\n",
    "1. `entity_dictionary` that contains the different tables resources that are available. The table we’re most interested in tends to be the participant table that contains the information about each participant.\n",
    "2. `data_dictionary` that contains the different field names that we might want to include in our dataset.\n",
    "3. `coding_dictionary` that contains a lookup for the values for some of the field names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Note: This cell can only be run once. Otherwise, you'll need to delete the existing data tables in order to re-run\n",
    "cmd = ['dx', 'extract_dataset', dataset, '-ddd', '--delimiter', ',']\n",
    "subprocess.check_call(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load cohorts that were created in cohort browser\n",
    "\n",
    "Cohorts were created in Cohort Browser. `ischemic_cases` cohort was created by having the following condition in the field [41270](https://biobank.ctsu.ox.ac.uk/crystal/field.cgi?id=41270) condition `I20-I25 Ischaemic heart disease`. `ischemic_controls` was created by using Cohorts compare: not in `ischemic_cases`.\n",
    "\n",
    "Here we use the `load_cohort` function from dxdata to load the cohort record. For more information about dxdata, see the associated [github repository](https://github.com/dnanexus/OpenBio/blob/master/dxdata/getting_started_with_dxdata.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Discover cohort data\n",
    "dispensed_control_id = list(\n",
    "    dxpy.find_data_objects(\n",
    "        typename='CohortBrowser',\n",
    "        folder='/Cohorts',\n",
    "        name_mode='exact',\n",
    "        name='ischemic_controls',\n",
    "    )\n",
    ")[0]['id']\n",
    "\n",
    "dispensed_case_id = list(\n",
    "    dxpy.find_data_objects(\n",
    "        typename='CohortBrowser',\n",
    "        folder='/Cohorts',\n",
    "        name_mode='exact',\n",
    "        name='ischemic_cases',\n",
    "    )\n",
    ")[0]['id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "control_dataset = (':').join([project_id, dispensed_control_id])\n",
    "case_dataset = (':').join([project_id, dispensed_case_id])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Retrieve phenotypic data\n",
    "\n",
    "Specify fields ID to retrieve, get corresponding UKB RAP field names and print description table.\n",
    "\n",
    "- `31` - [Sex](https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=31)\n",
    "- `2966` - [Age high blood pressure diagnosed](https://biobank.ctsu.ox.ac.uk/crystal/field.cgi?id=2966)\n",
    "- `22001` - [Genetic sex](https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=22001)\n",
    "- `22006` - [Genetic ethnic grouping](https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=22006)\n",
    "- `22019` - [Sex chromosome aneuploidy](https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=22019)\n",
    "- `22021` - [Genetic kinship to other participants](https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=22021)\n",
    "- `21022` - [Age at recruitment](https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=21022)\n",
    "- `23104` - [Body mass index (BMI)](https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=23104)\n",
    "- `20160` - [Ever smoked](https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=20160)\n",
    "- `30760` - [HDL cholesterol](https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=30760)\n",
    "- `30780` - [LDL direct](https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=30780)\n",
    "- `22020` - [Used in genetic principal components](https://biobank.ndph.ox.ac.uk/ukb/field.cgi?id=22020)\n",
    "- `22009` - [Genetic principal components](https://biobank.ndph.ox.ac.uk/showcase/field.cgi?id=22009)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "field_ids = [\n",
    "    '31',\n",
    "    '2966',\n",
    "    '22001',\n",
    "    '22006',\n",
    "    '22019',\n",
    "    '22021',\n",
    "    '21022',\n",
    "    '23104',\n",
    "    '20160',\n",
    "    '30760',\n",
    "    '30780',\n",
    "    '22020',\n",
    "    '22009'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_978/2068367959.py:2: DtypeWarning: Columns (4,7,8,9,10,15) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data_dict_df = pd.read_csv(data_dict_csv)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entity</th>\n",
       "      <th>name</th>\n",
       "      <th>type</th>\n",
       "      <th>primary_key_type</th>\n",
       "      <th>coding_name</th>\n",
       "      <th>concept</th>\n",
       "      <th>description</th>\n",
       "      <th>folder_path</th>\n",
       "      <th>is_multi_select</th>\n",
       "      <th>is_sparse_coding</th>\n",
       "      <th>linkout</th>\n",
       "      <th>longitudinal_axis_type</th>\n",
       "      <th>referenced_entity_field</th>\n",
       "      <th>relationship</th>\n",
       "      <th>title</th>\n",
       "      <th>units</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>participant</td>\n",
       "      <td>eid</td>\n",
       "      <td>string</td>\n",
       "      <td>global</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Participant Information</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Participant ID</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>participant</td>\n",
       "      <td>p3_i0</td>\n",
       "      <td>integer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Assessment centre &gt; Procedural metrics &gt; Proce...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://biobank.ctsu.ox.ac.uk/crystal/field.cgi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Verbal interview duration | Instance 0</td>\n",
       "      <td>seconds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>participant</td>\n",
       "      <td>p3_i1</td>\n",
       "      <td>integer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Assessment centre &gt; Procedural metrics &gt; Proce...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://biobank.ctsu.ox.ac.uk/crystal/field.cgi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Verbal interview duration | Instance 1</td>\n",
       "      <td>seconds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>participant</td>\n",
       "      <td>p3_i2</td>\n",
       "      <td>integer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Assessment centre &gt; Procedural metrics &gt; Proce...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://biobank.ctsu.ox.ac.uk/crystal/field.cgi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Verbal interview duration | Instance 2</td>\n",
       "      <td>seconds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>participant</td>\n",
       "      <td>p3_i3</td>\n",
       "      <td>integer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Assessment centre &gt; Procedural metrics &gt; Proce...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://biobank.ctsu.ox.ac.uk/crystal/field.cgi...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Verbal interview duration | Instance 3</td>\n",
       "      <td>seconds</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        entity   name     type primary_key_type coding_name  concept  \\\n",
       "0  participant    eid   string           global         NaN      NaN   \n",
       "1  participant  p3_i0  integer              NaN         NaN      NaN   \n",
       "2  participant  p3_i1  integer              NaN         NaN      NaN   \n",
       "3  participant  p3_i2  integer              NaN         NaN      NaN   \n",
       "4  participant  p3_i3  integer              NaN         NaN      NaN   \n",
       "\n",
       "   description                                        folder_path  \\\n",
       "0          NaN                            Participant Information   \n",
       "1          NaN  Assessment centre > Procedural metrics > Proce...   \n",
       "2          NaN  Assessment centre > Procedural metrics > Proce...   \n",
       "3          NaN  Assessment centre > Procedural metrics > Proce...   \n",
       "4          NaN  Assessment centre > Procedural metrics > Proce...   \n",
       "\n",
       "  is_multi_select is_sparse_coding  \\\n",
       "0             NaN              NaN   \n",
       "1             NaN              NaN   \n",
       "2             NaN              NaN   \n",
       "3             NaN              NaN   \n",
       "4             NaN              NaN   \n",
       "\n",
       "                                             linkout  longitudinal_axis_type  \\\n",
       "0                                                NaN                     NaN   \n",
       "1  http://biobank.ctsu.ox.ac.uk/crystal/field.cgi...                     NaN   \n",
       "2  http://biobank.ctsu.ox.ac.uk/crystal/field.cgi...                     NaN   \n",
       "3  http://biobank.ctsu.ox.ac.uk/crystal/field.cgi...                     NaN   \n",
       "4  http://biobank.ctsu.ox.ac.uk/crystal/field.cgi...                     NaN   \n",
       "\n",
       "  referenced_entity_field relationship  \\\n",
       "0                     NaN          NaN   \n",
       "1                     NaN          NaN   \n",
       "2                     NaN          NaN   \n",
       "3                     NaN          NaN   \n",
       "4                     NaN          NaN   \n",
       "\n",
       "                                    title    units  \n",
       "0                          Participant ID      NaN  \n",
       "1  Verbal interview duration | Instance 0  seconds  \n",
       "2  Verbal interview duration | Instance 1  seconds  \n",
       "3  Verbal interview duration | Instance 2  seconds  \n",
       "4  Verbal interview duration | Instance 3  seconds  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_dict_csv = glob.glob(os.path.join(path, '*.data_dictionary.csv'))[0]\n",
    "data_dict_df = pd.read_csv(data_dict_csv)\n",
    "data_dict_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The UKB participant tables have the following naming convention for the fields (or columns): `p<field id>_i<instance id>_a<array id>`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fields_for_id(field_id):\n",
    "    '''Collect all field names (e.g. 'p<field_id>_iYYY_aZZZ') given a list of field IDs and return string to pass into extract_dataset'''\n",
    "    field_names = ['eid']\n",
    "    for _id in field_id:\n",
    "        select_field_names = list(\n",
    "            data_dict_df[\n",
    "                data_dict_df.name.str.match(r'^p{}(_i\\d+)?(_a\\d+)?$'.format(_id))\n",
    "            ].name.values\n",
    "        )\n",
    "        # Note: This conditional is used to select only the first instance for all fields except '2966'\n",
    "        # This conditional is not needed otherwise\n",
    "        # For PCA field, select the first ten PCs\n",
    "        if _id == '22009':\n",
    "            field_names += select_field_names[:10]\n",
    "        # Select only the first instance for all fields except '2966' \n",
    "        elif _id != '2966' and len(select_field_names) > 1:\n",
    "            field_names.append(select_field_names[0])\n",
    "        else:\n",
    "            field_names += select_field_names\n",
    "\n",
    "    field_names = [f'participant.{f}' for f in field_names]\n",
    "    return ','.join(field_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'participant.eid,participant.p31,participant.p2966_i0,participant.p2966_i1,participant.p2966_i2,participant.p2966_i3,participant.p22001,participant.p22006,participant.p22019,participant.p22021,participant.p21022,participant.p23104_i0,participant.p20160_i0,participant.p30760_i0,participant.p30780_i0,participant.p22020,participant.p22009_a1,participant.p22009_a2,participant.p22009_a3,participant.p22009_a4,participant.p22009_a5,participant.p22009_a6,participant.p22009_a7,participant.p22009_a8,participant.p22009_a9,participant.p22009_a10'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "field_names = fields_for_id(field_ids)\n",
    "field_names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select phenotypes for case and control cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load dataset\n",
    "# Note: This cell can only be run once. Otherwise, you'll need to delete the existing data tables in order to re-run\n",
    "# Note: There is no space separating the different fields\n",
    "cmd = [\n",
    "    'dx',\n",
    "    'extract_dataset',\n",
    "    control_dataset,\n",
    "    '--fields',\n",
    "    field_names,\n",
    "    '--delimiter',\n",
    "    ',',\n",
    "    '--output',\n",
    "    'control_dictionary.csv',\n",
    "]\n",
    "subprocess.check_call(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmd = [\n",
    "    'dx',\n",
    "    'extract_dataset',\n",
    "    case_dataset,\n",
    "    '--fields',\n",
    "    field_names,\n",
    "    '--delimiter',\n",
    "    ',',\n",
    "    '--output',\n",
    "    'case_dictionary.csv',\n",
    "]\n",
    "subprocess.check_call(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(400946, 26)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant.eid</th>\n",
       "      <th>participant.p31</th>\n",
       "      <th>participant.p2966_i0</th>\n",
       "      <th>participant.p2966_i1</th>\n",
       "      <th>participant.p2966_i2</th>\n",
       "      <th>participant.p2966_i3</th>\n",
       "      <th>participant.p22001</th>\n",
       "      <th>participant.p22006</th>\n",
       "      <th>participant.p22019</th>\n",
       "      <th>participant.p22021</th>\n",
       "      <th>...</th>\n",
       "      <th>participant.p22009_a1</th>\n",
       "      <th>participant.p22009_a2</th>\n",
       "      <th>participant.p22009_a3</th>\n",
       "      <th>participant.p22009_a4</th>\n",
       "      <th>participant.p22009_a5</th>\n",
       "      <th>participant.p22009_a6</th>\n",
       "      <th>participant.p22009_a7</th>\n",
       "      <th>participant.p22009_a8</th>\n",
       "      <th>participant.p22009_a9</th>\n",
       "      <th>participant.p22009_a10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000128</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-12.10990</td>\n",
       "      <td>3.53791</td>\n",
       "      <td>-0.584297</td>\n",
       "      <td>0.604633</td>\n",
       "      <td>-2.46493</td>\n",
       "      <td>-0.448158</td>\n",
       "      <td>3.393600</td>\n",
       "      <td>2.100990</td>\n",
       "      <td>-11.98610</td>\n",
       "      <td>-3.00294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000424</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.24265</td>\n",
       "      <td>1.92411</td>\n",
       "      <td>0.169679</td>\n",
       "      <td>3.306660</td>\n",
       "      <td>-4.06993</td>\n",
       "      <td>-2.712060</td>\n",
       "      <td>2.389620</td>\n",
       "      <td>-1.731740</td>\n",
       "      <td>2.02853</td>\n",
       "      <td>1.68914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000569</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-11.48970</td>\n",
       "      <td>3.69699</td>\n",
       "      <td>-3.154340</td>\n",
       "      <td>-0.964846</td>\n",
       "      <td>-6.10410</td>\n",
       "      <td>-2.444060</td>\n",
       "      <td>-1.613400</td>\n",
       "      <td>-1.357070</td>\n",
       "      <td>4.39817</td>\n",
       "      <td>1.29341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1000695</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-11.55390</td>\n",
       "      <td>1.22231</td>\n",
       "      <td>-1.271690</td>\n",
       "      <td>0.461033</td>\n",
       "      <td>-4.63990</td>\n",
       "      <td>1.703520</td>\n",
       "      <td>0.433453</td>\n",
       "      <td>-0.659432</td>\n",
       "      <td>1.15092</td>\n",
       "      <td>2.71503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1000747</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.26790</td>\n",
       "      <td>2.05402</td>\n",
       "      <td>-1.896930</td>\n",
       "      <td>3.913610</td>\n",
       "      <td>6.11443</td>\n",
       "      <td>0.617679</td>\n",
       "      <td>0.177710</td>\n",
       "      <td>-0.669613</td>\n",
       "      <td>1.70910</td>\n",
       "      <td>-1.61138</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   participant.eid  participant.p31  participant.p2966_i0  \\\n",
       "0          1000128                1                   NaN   \n",
       "1          1000424                1                   NaN   \n",
       "2          1000569                1                   NaN   \n",
       "3          1000695                1                   NaN   \n",
       "4          1000747                0                   NaN   \n",
       "\n",
       "   participant.p2966_i1  participant.p2966_i2  participant.p2966_i3  \\\n",
       "0                   NaN                   NaN                   NaN   \n",
       "1                   NaN                   NaN                   NaN   \n",
       "2                   NaN                   NaN                   NaN   \n",
       "3                   NaN                   NaN                   NaN   \n",
       "4                   NaN                   NaN                   NaN   \n",
       "\n",
       "   participant.p22001  participant.p22006  participant.p22019  \\\n",
       "0                 1.0                 1.0                 NaN   \n",
       "1                 1.0                 1.0                 NaN   \n",
       "2                 1.0                 1.0                 NaN   \n",
       "3                 1.0                 1.0                 NaN   \n",
       "4                 0.0                 1.0                 NaN   \n",
       "\n",
       "   participant.p22021  ...  participant.p22009_a1  participant.p22009_a2  \\\n",
       "0                 0.0  ...              -12.10990                3.53791   \n",
       "1                 0.0  ...               -9.24265                1.92411   \n",
       "2                 0.0  ...              -11.48970                3.69699   \n",
       "3                 0.0  ...              -11.55390                1.22231   \n",
       "4                 0.0  ...              -13.26790                2.05402   \n",
       "\n",
       "   participant.p22009_a3  participant.p22009_a4  participant.p22009_a5  \\\n",
       "0              -0.584297               0.604633               -2.46493   \n",
       "1               0.169679               3.306660               -4.06993   \n",
       "2              -3.154340              -0.964846               -6.10410   \n",
       "3              -1.271690               0.461033               -4.63990   \n",
       "4              -1.896930               3.913610                6.11443   \n",
       "\n",
       "   participant.p22009_a6  participant.p22009_a7  participant.p22009_a8  \\\n",
       "0              -0.448158               3.393600               2.100990   \n",
       "1              -2.712060               2.389620              -1.731740   \n",
       "2              -2.444060              -1.613400              -1.357070   \n",
       "3               1.703520               0.433453              -0.659432   \n",
       "4               0.617679               0.177710              -0.669613   \n",
       "\n",
       "   participant.p22009_a9  participant.p22009_a10  \n",
       "0              -11.98610                -3.00294  \n",
       "1                2.02853                 1.68914  \n",
       "2                4.39817                 1.29341  \n",
       "3                1.15092                 2.71503  \n",
       "4                1.70910                -1.61138  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_dict_csv = 'control_dictionary.csv'\n",
    "control_df = pd.read_csv(control_dict_csv)\n",
    "print(control_df.shape)\n",
    "control_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eid</th>\n",
       "      <th>p31</th>\n",
       "      <th>p2966_i0</th>\n",
       "      <th>p2966_i1</th>\n",
       "      <th>p2966_i2</th>\n",
       "      <th>p2966_i3</th>\n",
       "      <th>p22001</th>\n",
       "      <th>p22006</th>\n",
       "      <th>p22019</th>\n",
       "      <th>p22021</th>\n",
       "      <th>...</th>\n",
       "      <th>p22009_a1</th>\n",
       "      <th>p22009_a2</th>\n",
       "      <th>p22009_a3</th>\n",
       "      <th>p22009_a4</th>\n",
       "      <th>p22009_a5</th>\n",
       "      <th>p22009_a6</th>\n",
       "      <th>p22009_a7</th>\n",
       "      <th>p22009_a8</th>\n",
       "      <th>p22009_a9</th>\n",
       "      <th>p22009_a10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000128</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-12.10990</td>\n",
       "      <td>3.53791</td>\n",
       "      <td>-0.584297</td>\n",
       "      <td>0.604633</td>\n",
       "      <td>-2.46493</td>\n",
       "      <td>-0.448158</td>\n",
       "      <td>3.393600</td>\n",
       "      <td>2.100990</td>\n",
       "      <td>-11.98610</td>\n",
       "      <td>-3.00294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000424</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.24265</td>\n",
       "      <td>1.92411</td>\n",
       "      <td>0.169679</td>\n",
       "      <td>3.306660</td>\n",
       "      <td>-4.06993</td>\n",
       "      <td>-2.712060</td>\n",
       "      <td>2.389620</td>\n",
       "      <td>-1.731740</td>\n",
       "      <td>2.02853</td>\n",
       "      <td>1.68914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000569</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-11.48970</td>\n",
       "      <td>3.69699</td>\n",
       "      <td>-3.154340</td>\n",
       "      <td>-0.964846</td>\n",
       "      <td>-6.10410</td>\n",
       "      <td>-2.444060</td>\n",
       "      <td>-1.613400</td>\n",
       "      <td>-1.357070</td>\n",
       "      <td>4.39817</td>\n",
       "      <td>1.29341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1000695</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-11.55390</td>\n",
       "      <td>1.22231</td>\n",
       "      <td>-1.271690</td>\n",
       "      <td>0.461033</td>\n",
       "      <td>-4.63990</td>\n",
       "      <td>1.703520</td>\n",
       "      <td>0.433453</td>\n",
       "      <td>-0.659432</td>\n",
       "      <td>1.15092</td>\n",
       "      <td>2.71503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1000747</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.26790</td>\n",
       "      <td>2.05402</td>\n",
       "      <td>-1.896930</td>\n",
       "      <td>3.913610</td>\n",
       "      <td>6.11443</td>\n",
       "      <td>0.617679</td>\n",
       "      <td>0.177710</td>\n",
       "      <td>-0.669613</td>\n",
       "      <td>1.70910</td>\n",
       "      <td>-1.61138</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       eid  p31  p2966_i0  p2966_i1  p2966_i2  p2966_i3  p22001  p22006  \\\n",
       "0  1000128    1       NaN       NaN       NaN       NaN     1.0     1.0   \n",
       "1  1000424    1       NaN       NaN       NaN       NaN     1.0     1.0   \n",
       "2  1000569    1       NaN       NaN       NaN       NaN     1.0     1.0   \n",
       "3  1000695    1       NaN       NaN       NaN       NaN     1.0     1.0   \n",
       "4  1000747    0       NaN       NaN       NaN       NaN     0.0     1.0   \n",
       "\n",
       "   p22019  p22021  ...  p22009_a1  p22009_a2  p22009_a3  p22009_a4  p22009_a5  \\\n",
       "0     NaN     0.0  ...  -12.10990    3.53791  -0.584297   0.604633   -2.46493   \n",
       "1     NaN     0.0  ...   -9.24265    1.92411   0.169679   3.306660   -4.06993   \n",
       "2     NaN     0.0  ...  -11.48970    3.69699  -3.154340  -0.964846   -6.10410   \n",
       "3     NaN     0.0  ...  -11.55390    1.22231  -1.271690   0.461033   -4.63990   \n",
       "4     NaN     0.0  ...  -13.26790    2.05402  -1.896930   3.913610    6.11443   \n",
       "\n",
       "   p22009_a6  p22009_a7  p22009_a8  p22009_a9  p22009_a10  \n",
       "0  -0.448158   3.393600   2.100990  -11.98610    -3.00294  \n",
       "1  -2.712060   2.389620  -1.731740    2.02853     1.68914  \n",
       "2  -2.444060  -1.613400  -1.357070    4.39817     1.29341  \n",
       "3   1.703520   0.433453  -0.659432    1.15092     2.71503  \n",
       "4   0.617679   0.177710  -0.669613    1.70910    -1.61138  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename column headers\n",
    "control_df = control_df.rename(columns=lambda x: re.sub('participant.', '', x))\n",
    "control_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(91331, 26)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant.eid</th>\n",
       "      <th>participant.p31</th>\n",
       "      <th>participant.p2966_i0</th>\n",
       "      <th>participant.p2966_i1</th>\n",
       "      <th>participant.p2966_i2</th>\n",
       "      <th>participant.p2966_i3</th>\n",
       "      <th>participant.p22001</th>\n",
       "      <th>participant.p22006</th>\n",
       "      <th>participant.p22019</th>\n",
       "      <th>participant.p22021</th>\n",
       "      <th>...</th>\n",
       "      <th>participant.p22009_a1</th>\n",
       "      <th>participant.p22009_a2</th>\n",
       "      <th>participant.p22009_a3</th>\n",
       "      <th>participant.p22009_a4</th>\n",
       "      <th>participant.p22009_a5</th>\n",
       "      <th>participant.p22009_a6</th>\n",
       "      <th>participant.p22009_a7</th>\n",
       "      <th>participant.p22009_a8</th>\n",
       "      <th>participant.p22009_a9</th>\n",
       "      <th>participant.p22009_a10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000639</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>21.30840</td>\n",
       "      <td>-14.41990</td>\n",
       "      <td>36.33220</td>\n",
       "      <td>-68.67570</td>\n",
       "      <td>6.79464</td>\n",
       "      <td>2.918310</td>\n",
       "      <td>34.677700</td>\n",
       "      <td>20.404300</td>\n",
       "      <td>0.293947</td>\n",
       "      <td>1.722900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1001142</td>\n",
       "      <td>0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.76094</td>\n",
       "      <td>-10.55370</td>\n",
       "      <td>-6.37599</td>\n",
       "      <td>-2.46630</td>\n",
       "      <td>-5.08631</td>\n",
       "      <td>-3.109580</td>\n",
       "      <td>4.386370</td>\n",
       "      <td>3.252660</td>\n",
       "      <td>-3.096380</td>\n",
       "      <td>-2.796150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1001251</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-14.24570</td>\n",
       "      <td>3.96257</td>\n",
       "      <td>-1.01864</td>\n",
       "      <td>2.27742</td>\n",
       "      <td>-4.21549</td>\n",
       "      <td>-0.100448</td>\n",
       "      <td>1.524450</td>\n",
       "      <td>-1.246880</td>\n",
       "      <td>1.368960</td>\n",
       "      <td>2.570260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1001445</td>\n",
       "      <td>1</td>\n",
       "      <td>60.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.43690</td>\n",
       "      <td>3.02070</td>\n",
       "      <td>-3.19577</td>\n",
       "      <td>0.42849</td>\n",
       "      <td>6.21226</td>\n",
       "      <td>-1.478700</td>\n",
       "      <td>1.282360</td>\n",
       "      <td>-1.606150</td>\n",
       "      <td>0.199501</td>\n",
       "      <td>-4.308460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1002533</td>\n",
       "      <td>0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.81500</td>\n",
       "      <td>2.60585</td>\n",
       "      <td>-5.53910</td>\n",
       "      <td>8.82807</td>\n",
       "      <td>22.86970</td>\n",
       "      <td>2.332280</td>\n",
       "      <td>0.505821</td>\n",
       "      <td>0.470971</td>\n",
       "      <td>1.502100</td>\n",
       "      <td>0.467284</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   participant.eid  participant.p31  participant.p2966_i0  \\\n",
       "0          1000639                1                   NaN   \n",
       "1          1001142                0                  55.0   \n",
       "2          1001251                0                   NaN   \n",
       "3          1001445                1                  60.0   \n",
       "4          1002533                0                  40.0   \n",
       "\n",
       "   participant.p2966_i1  participant.p2966_i2  participant.p2966_i3  \\\n",
       "0                   NaN                   NaN                   NaN   \n",
       "1                   NaN                   NaN                   NaN   \n",
       "2                   NaN                   NaN                   NaN   \n",
       "3                   NaN                   NaN                   NaN   \n",
       "4                   NaN                   NaN                   NaN   \n",
       "\n",
       "   participant.p22001  participant.p22006  participant.p22019  \\\n",
       "0                 1.0                 NaN                 NaN   \n",
       "1                 0.0                 NaN                 NaN   \n",
       "2                 0.0                 1.0                 NaN   \n",
       "3                 1.0                 1.0                 NaN   \n",
       "4                 0.0                 NaN                 NaN   \n",
       "\n",
       "   participant.p22021  ...  participant.p22009_a1  participant.p22009_a2  \\\n",
       "0                 0.0  ...               21.30840              -14.41990   \n",
       "1                 0.0  ...               -1.76094              -10.55370   \n",
       "2                 0.0  ...              -14.24570                3.96257   \n",
       "3                 1.0  ...              -13.43690                3.02070   \n",
       "4                 0.0  ...              -13.81500                2.60585   \n",
       "\n",
       "   participant.p22009_a3  participant.p22009_a4  participant.p22009_a5  \\\n",
       "0               36.33220              -68.67570                6.79464   \n",
       "1               -6.37599               -2.46630               -5.08631   \n",
       "2               -1.01864                2.27742               -4.21549   \n",
       "3               -3.19577                0.42849                6.21226   \n",
       "4               -5.53910                8.82807               22.86970   \n",
       "\n",
       "   participant.p22009_a6  participant.p22009_a7  participant.p22009_a8  \\\n",
       "0               2.918310              34.677700              20.404300   \n",
       "1              -3.109580               4.386370               3.252660   \n",
       "2              -0.100448               1.524450              -1.246880   \n",
       "3              -1.478700               1.282360              -1.606150   \n",
       "4               2.332280               0.505821               0.470971   \n",
       "\n",
       "   participant.p22009_a9  participant.p22009_a10  \n",
       "0               0.293947                1.722900  \n",
       "1              -3.096380               -2.796150  \n",
       "2               1.368960                2.570260  \n",
       "3               0.199501               -4.308460  \n",
       "4               1.502100                0.467284  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case_dict_csv = 'case_dictionary.csv'\n",
    "case_df = pd.read_csv(case_dict_csv)\n",
    "print(case_df.shape)\n",
    "case_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eid</th>\n",
       "      <th>p31</th>\n",
       "      <th>p2966_i0</th>\n",
       "      <th>p2966_i1</th>\n",
       "      <th>p2966_i2</th>\n",
       "      <th>p2966_i3</th>\n",
       "      <th>p22001</th>\n",
       "      <th>p22006</th>\n",
       "      <th>p22019</th>\n",
       "      <th>p22021</th>\n",
       "      <th>...</th>\n",
       "      <th>p22009_a1</th>\n",
       "      <th>p22009_a2</th>\n",
       "      <th>p22009_a3</th>\n",
       "      <th>p22009_a4</th>\n",
       "      <th>p22009_a5</th>\n",
       "      <th>p22009_a6</th>\n",
       "      <th>p22009_a7</th>\n",
       "      <th>p22009_a8</th>\n",
       "      <th>p22009_a9</th>\n",
       "      <th>p22009_a10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000639</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>21.30840</td>\n",
       "      <td>-14.41990</td>\n",
       "      <td>36.33220</td>\n",
       "      <td>-68.67570</td>\n",
       "      <td>6.79464</td>\n",
       "      <td>2.918310</td>\n",
       "      <td>34.677700</td>\n",
       "      <td>20.404300</td>\n",
       "      <td>0.293947</td>\n",
       "      <td>1.722900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1001142</td>\n",
       "      <td>0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.76094</td>\n",
       "      <td>-10.55370</td>\n",
       "      <td>-6.37599</td>\n",
       "      <td>-2.46630</td>\n",
       "      <td>-5.08631</td>\n",
       "      <td>-3.109580</td>\n",
       "      <td>4.386370</td>\n",
       "      <td>3.252660</td>\n",
       "      <td>-3.096380</td>\n",
       "      <td>-2.796150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1001251</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-14.24570</td>\n",
       "      <td>3.96257</td>\n",
       "      <td>-1.01864</td>\n",
       "      <td>2.27742</td>\n",
       "      <td>-4.21549</td>\n",
       "      <td>-0.100448</td>\n",
       "      <td>1.524450</td>\n",
       "      <td>-1.246880</td>\n",
       "      <td>1.368960</td>\n",
       "      <td>2.570260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1001445</td>\n",
       "      <td>1</td>\n",
       "      <td>60.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.43690</td>\n",
       "      <td>3.02070</td>\n",
       "      <td>-3.19577</td>\n",
       "      <td>0.42849</td>\n",
       "      <td>6.21226</td>\n",
       "      <td>-1.478700</td>\n",
       "      <td>1.282360</td>\n",
       "      <td>-1.606150</td>\n",
       "      <td>0.199501</td>\n",
       "      <td>-4.308460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1002533</td>\n",
       "      <td>0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.81500</td>\n",
       "      <td>2.60585</td>\n",
       "      <td>-5.53910</td>\n",
       "      <td>8.82807</td>\n",
       "      <td>22.86970</td>\n",
       "      <td>2.332280</td>\n",
       "      <td>0.505821</td>\n",
       "      <td>0.470971</td>\n",
       "      <td>1.502100</td>\n",
       "      <td>0.467284</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       eid  p31  p2966_i0  p2966_i1  p2966_i2  p2966_i3  p22001  p22006  \\\n",
       "0  1000639    1       NaN       NaN       NaN       NaN     1.0     NaN   \n",
       "1  1001142    0      55.0       NaN       NaN       NaN     0.0     NaN   \n",
       "2  1001251    0       NaN       NaN       NaN       NaN     0.0     1.0   \n",
       "3  1001445    1      60.0       NaN       NaN       NaN     1.0     1.0   \n",
       "4  1002533    0      40.0       NaN       NaN       NaN     0.0     NaN   \n",
       "\n",
       "   p22019  p22021  ...  p22009_a1  p22009_a2  p22009_a3  p22009_a4  p22009_a5  \\\n",
       "0     NaN     0.0  ...   21.30840  -14.41990   36.33220  -68.67570    6.79464   \n",
       "1     NaN     0.0  ...   -1.76094  -10.55370   -6.37599   -2.46630   -5.08631   \n",
       "2     NaN     0.0  ...  -14.24570    3.96257   -1.01864    2.27742   -4.21549   \n",
       "3     NaN     1.0  ...  -13.43690    3.02070   -3.19577    0.42849    6.21226   \n",
       "4     NaN     0.0  ...  -13.81500    2.60585   -5.53910    8.82807   22.86970   \n",
       "\n",
       "   p22009_a6  p22009_a7  p22009_a8  p22009_a9  p22009_a10  \n",
       "0   2.918310  34.677700  20.404300   0.293947    1.722900  \n",
       "1  -3.109580   4.386370   3.252660  -3.096380   -2.796150  \n",
       "2  -0.100448   1.524450  -1.246880   1.368960    2.570260  \n",
       "3  -1.478700   1.282360  -1.606150   0.199501   -4.308460  \n",
       "4   2.332280   0.505821   0.470971   1.502100    0.467284  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case_df = case_df.rename(columns=lambda x: re.sub('participant.', '', x))\n",
    "case_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create phenotype variable and concatenate cohorts into one dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "case_df['ischemia_cc'] = 1\n",
    "control_df['ischemia_cc'] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.concat([case_df, control_df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    400946\n",
       "1     91331\n",
       "Name: ischemia_cc, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Note: The counts should be consistent with the counts from the Cohort Browser\n",
    "df.ischemia_cc.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is an example of retrieved data.\n",
    "\n",
    "*Note: this table contains synthetic data and is made for demonstration purposes only!*\n",
    "\n",
    "|Row|eid|p31|p2966_i0|p2966_i1|p2966_i2|p2966_i3|p22001|p22006|p22019|p22021|p22022|p23104_i0|p20160_i0|p30760_i0|p30780_i0|p22020|ischemia_cc|\n",
    "|:---|:--- |:---|:--- |:---|:---|:---|:---|:--- |:---|:---|:--- |:---|:--- |:---|:--- |:---|:---|\n",
    "|0 |EID-XXXXXXX|0 |58.0|NA|NA|NA|0.0|1.0|NA|0.0|61.0|28.5|1.0|0.959|2.456|1.0|1|\n",
    "|1 |EID-YYYYYYY|1 |NA  |NA|NA|NA|1.0|1.0|NA|0.0|51.0|24.5|1.0|1.159|2.719|1.0|1|\n",
    "|2 |EID-ZZZZZZZ|1 |50.0|NA|NA|NA|1.0|1.0|NA|1.0|64.0|29.7|1.0|0.970|3.440|NA |1|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(492277, 27)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eid</th>\n",
       "      <th>p31</th>\n",
       "      <th>p2966_i0</th>\n",
       "      <th>p2966_i1</th>\n",
       "      <th>p2966_i2</th>\n",
       "      <th>p2966_i3</th>\n",
       "      <th>p22001</th>\n",
       "      <th>p22006</th>\n",
       "      <th>p22019</th>\n",
       "      <th>p22021</th>\n",
       "      <th>...</th>\n",
       "      <th>p22009_a2</th>\n",
       "      <th>p22009_a3</th>\n",
       "      <th>p22009_a4</th>\n",
       "      <th>p22009_a5</th>\n",
       "      <th>p22009_a6</th>\n",
       "      <th>p22009_a7</th>\n",
       "      <th>p22009_a8</th>\n",
       "      <th>p22009_a9</th>\n",
       "      <th>p22009_a10</th>\n",
       "      <th>ischemia_cc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000639</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-14.41990</td>\n",
       "      <td>36.33220</td>\n",
       "      <td>-68.67570</td>\n",
       "      <td>6.79464</td>\n",
       "      <td>2.918310</td>\n",
       "      <td>34.677700</td>\n",
       "      <td>20.404300</td>\n",
       "      <td>0.293947</td>\n",
       "      <td>1.722900</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1001142</td>\n",
       "      <td>0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-10.55370</td>\n",
       "      <td>-6.37599</td>\n",
       "      <td>-2.46630</td>\n",
       "      <td>-5.08631</td>\n",
       "      <td>-3.109580</td>\n",
       "      <td>4.386370</td>\n",
       "      <td>3.252660</td>\n",
       "      <td>-3.096380</td>\n",
       "      <td>-2.796150</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1001251</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.96257</td>\n",
       "      <td>-1.01864</td>\n",
       "      <td>2.27742</td>\n",
       "      <td>-4.21549</td>\n",
       "      <td>-0.100448</td>\n",
       "      <td>1.524450</td>\n",
       "      <td>-1.246880</td>\n",
       "      <td>1.368960</td>\n",
       "      <td>2.570260</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1001445</td>\n",
       "      <td>1</td>\n",
       "      <td>60.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.02070</td>\n",
       "      <td>-3.19577</td>\n",
       "      <td>0.42849</td>\n",
       "      <td>6.21226</td>\n",
       "      <td>-1.478700</td>\n",
       "      <td>1.282360</td>\n",
       "      <td>-1.606150</td>\n",
       "      <td>0.199501</td>\n",
       "      <td>-4.308460</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1002533</td>\n",
       "      <td>0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.60585</td>\n",
       "      <td>-5.53910</td>\n",
       "      <td>8.82807</td>\n",
       "      <td>22.86970</td>\n",
       "      <td>2.332280</td>\n",
       "      <td>0.505821</td>\n",
       "      <td>0.470971</td>\n",
       "      <td>1.502100</td>\n",
       "      <td>0.467284</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       eid  p31  p2966_i0  p2966_i1  p2966_i2  p2966_i3  p22001  p22006  \\\n",
       "0  1000639    1       NaN       NaN       NaN       NaN     1.0     NaN   \n",
       "1  1001142    0      55.0       NaN       NaN       NaN     0.0     NaN   \n",
       "2  1001251    0       NaN       NaN       NaN       NaN     0.0     1.0   \n",
       "3  1001445    1      60.0       NaN       NaN       NaN     1.0     1.0   \n",
       "4  1002533    0      40.0       NaN       NaN       NaN     0.0     NaN   \n",
       "\n",
       "   p22019  p22021  ...  p22009_a2  p22009_a3  p22009_a4  p22009_a5  p22009_a6  \\\n",
       "0     NaN     0.0  ...  -14.41990   36.33220  -68.67570    6.79464   2.918310   \n",
       "1     NaN     0.0  ...  -10.55370   -6.37599   -2.46630   -5.08631  -3.109580   \n",
       "2     NaN     0.0  ...    3.96257   -1.01864    2.27742   -4.21549  -0.100448   \n",
       "3     NaN     1.0  ...    3.02070   -3.19577    0.42849    6.21226  -1.478700   \n",
       "4     NaN     0.0  ...    2.60585   -5.53910    8.82807   22.86970   2.332280   \n",
       "\n",
       "   p22009_a7  p22009_a8  p22009_a9  p22009_a10  ischemia_cc  \n",
       "0  34.677700  20.404300   0.293947    1.722900            1  \n",
       "1   4.386370   3.252660  -3.096380   -2.796150            1  \n",
       "2   1.524450  -1.246880   1.368960    2.570260            1  \n",
       "3   1.282360  -1.606150   0.199501   -4.308460            1  \n",
       "4   0.505821   0.470971   1.502100    0.467284            1  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## QC samples based on several conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_qced = df[\n",
    "    (df['p31'] == df['p22001']) # Filter reported sex and genetic sex are the same\n",
    "    & (df['p22006'] == 1)  # Only include Caucasian ancestry (In_white_british_ancestry_subset)\n",
    "    & (  \n",
    "        df['p22019'].isnull() # No Sex chromosome aneuploidy\n",
    "    )\n",
    "    & (  \n",
    "        df['p22020'] == 1 # Participant was used to calculate PCA (only non-relatives were included)\n",
    "    )  \n",
    "].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    273298\n",
       "1     58698\n",
       "Name: ischemia_cc, dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_qced.ischemia_cc.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eid</th>\n",
       "      <th>p31</th>\n",
       "      <th>p2966_i0</th>\n",
       "      <th>p2966_i1</th>\n",
       "      <th>p2966_i2</th>\n",
       "      <th>p2966_i3</th>\n",
       "      <th>p22001</th>\n",
       "      <th>p22006</th>\n",
       "      <th>p22019</th>\n",
       "      <th>p22021</th>\n",
       "      <th>...</th>\n",
       "      <th>p22009_a2</th>\n",
       "      <th>p22009_a3</th>\n",
       "      <th>p22009_a4</th>\n",
       "      <th>p22009_a5</th>\n",
       "      <th>p22009_a6</th>\n",
       "      <th>p22009_a7</th>\n",
       "      <th>p22009_a8</th>\n",
       "      <th>p22009_a9</th>\n",
       "      <th>p22009_a10</th>\n",
       "      <th>ischemia_cc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1001251</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.96257</td>\n",
       "      <td>-1.018640</td>\n",
       "      <td>2.277420</td>\n",
       "      <td>-4.21549</td>\n",
       "      <td>-0.100448</td>\n",
       "      <td>1.524450</td>\n",
       "      <td>-1.246880</td>\n",
       "      <td>1.368960</td>\n",
       "      <td>2.570260</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1001445</td>\n",
       "      <td>1</td>\n",
       "      <td>60.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.02070</td>\n",
       "      <td>-3.195770</td>\n",
       "      <td>0.428490</td>\n",
       "      <td>6.21226</td>\n",
       "      <td>-1.478700</td>\n",
       "      <td>1.282360</td>\n",
       "      <td>-1.606150</td>\n",
       "      <td>0.199501</td>\n",
       "      <td>-4.308460</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1003040</td>\n",
       "      <td>0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.63916</td>\n",
       "      <td>-2.339450</td>\n",
       "      <td>0.502712</td>\n",
       "      <td>-2.63828</td>\n",
       "      <td>0.662364</td>\n",
       "      <td>0.931805</td>\n",
       "      <td>0.987020</td>\n",
       "      <td>-5.069820</td>\n",
       "      <td>2.115390</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1004201</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.71645</td>\n",
       "      <td>-0.737745</td>\n",
       "      <td>1.572150</td>\n",
       "      <td>-1.20120</td>\n",
       "      <td>-1.681440</td>\n",
       "      <td>1.290020</td>\n",
       "      <td>-0.223860</td>\n",
       "      <td>-5.693960</td>\n",
       "      <td>-1.878020</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1004219</td>\n",
       "      <td>0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.01404</td>\n",
       "      <td>-2.064440</td>\n",
       "      <td>1.280780</td>\n",
       "      <td>-1.68080</td>\n",
       "      <td>2.864480</td>\n",
       "      <td>2.142890</td>\n",
       "      <td>-0.380478</td>\n",
       "      <td>-4.603690</td>\n",
       "      <td>0.058457</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       eid  p31  p2966_i0  p2966_i1  p2966_i2  p2966_i3  p22001  p22006  \\\n",
       "2  1001251    0       NaN       NaN       NaN       NaN     0.0     1.0   \n",
       "3  1001445    1      60.0       NaN       NaN       NaN     1.0     1.0   \n",
       "5  1003040    0      54.0       NaN       NaN       NaN     0.0     1.0   \n",
       "8  1004201    0       NaN       NaN       NaN       NaN     0.0     1.0   \n",
       "9  1004219    0      28.0       NaN       NaN       NaN     0.0     1.0   \n",
       "\n",
       "   p22019  p22021  ...  p22009_a2  p22009_a3  p22009_a4  p22009_a5  p22009_a6  \\\n",
       "2     NaN     0.0  ...    3.96257  -1.018640   2.277420   -4.21549  -0.100448   \n",
       "3     NaN     1.0  ...    3.02070  -3.195770   0.428490    6.21226  -1.478700   \n",
       "5     NaN     1.0  ...    4.63916  -2.339450   0.502712   -2.63828   0.662364   \n",
       "8     NaN     1.0  ...    3.71645  -0.737745   1.572150   -1.20120  -1.681440   \n",
       "9     NaN     0.0  ...    2.01404  -2.064440   1.280780   -1.68080   2.864480   \n",
       "\n",
       "   p22009_a7  p22009_a8  p22009_a9  p22009_a10  ischemia_cc  \n",
       "2   1.524450  -1.246880   1.368960    2.570260            1  \n",
       "3   1.282360  -1.606150   0.199501   -4.308460            1  \n",
       "5   0.931805   0.987020  -5.069820    2.115390            1  \n",
       "8   1.290020  -0.223860  -5.693960   -1.878020            1  \n",
       "9   2.142890  -0.380478  -4.603690    0.058457            1  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_qced.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare covariates for subsequent regression analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Combine all instances of `Age high blood pressure diagnosed` (`2966`) field and make `hypertension` boolean variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_qced['hypertension'] = [\n",
    "    np.nansum([float(row[i]) for i in [0, 1, 2, 3]]) > 0\n",
    "    for row in df_qced[['p2966_i0', 'p2966_i1', 'p2966_i2', 'p2966_i3']].to_numpy()\n",
    "]\n",
    "df_qced['hypertension'] = df_qced['hypertension'].astype(int)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the `Ever smoked` (`20160`) make all missing values as `0`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_qced['p20160_i0'].fillna(0, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the rest of covariate columns, replace missing values by mean for that column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_qced['p23104_i0'].fillna(df_qced['p23104_i0'].mean(), inplace=True)\n",
    "df_qced['p30760_i0'].fillna(df_qced['p30760_i0'].mean(), inplace=True)\n",
    "df_qced['p30780_i0'].fillna(df_qced['p30780_i0'].mean(), inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "eid                  0\n",
       "p31                  0\n",
       "p2966_i0        243046\n",
       "p2966_i1        327766\n",
       "p2966_i2        317869\n",
       "p2966_i3        330814\n",
       "p22001               0\n",
       "p22006               0\n",
       "p22019          331996\n",
       "p22021               0\n",
       "p21022               0\n",
       "p23104_i0            0\n",
       "p20160_i0            0\n",
       "p30760_i0            0\n",
       "p30780_i0            0\n",
       "p22020               0\n",
       "p22009_a1            0\n",
       "p22009_a2            0\n",
       "p22009_a3            0\n",
       "p22009_a4            0\n",
       "p22009_a5            0\n",
       "p22009_a6            0\n",
       "p22009_a7            0\n",
       "p22009_a8            0\n",
       "p22009_a9            0\n",
       "p22009_a10           0\n",
       "ischemia_cc          0\n",
       "hypertension         0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_qced.isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rename columns and organize it in format suitable for PLINK and regenie"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename columns for better readibility\n",
    "df_qced = df_qced.rename(columns=lambda x: re.sub('p22009_a','pc',x))\n",
    "df_qced = df_qced.rename(\n",
    "    columns={\n",
    "        'eid': 'IID',\n",
    "        'p31': 'sex',\n",
    "        'p21022': 'age',\n",
    "        'p20160_i0': 'ever_smoked',\n",
    "        'p23104_i0': 'bmi',\n",
    "        'p30760_i0': 'hdl_cholesterol',\n",
    "        'p30780_i0': 'ldl_cholesterol',\n",
    "    }\n",
    ")\n",
    "\n",
    "# Add FID column -- required input format for regenie\n",
    "df_qced['FID'] = df_qced['IID']\n",
    "\n",
    "# Create a phenotype table from our QCed data\n",
    "cols = [\n",
    "        'FID',\n",
    "        'IID',\n",
    "        'sex',\n",
    "        'age',\n",
    "        'bmi',\n",
    "        'ever_smoked',\n",
    "        'hdl_cholesterol',\n",
    "        'ldl_cholesterol',\n",
    "        'hypertension',\n",
    "        'ischemia_cc',\n",
    "]\n",
    "cols.extend([col for col in df_qced if 'pc' in col])\n",
    "df_phenotype = df_qced[cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>bmi</th>\n",
       "      <th>ever_smoked</th>\n",
       "      <th>hdl_cholesterol</th>\n",
       "      <th>ldl_cholesterol</th>\n",
       "      <th>hypertension</th>\n",
       "      <th>ischemia_cc</th>\n",
       "      <th>pc1</th>\n",
       "      <th>pc2</th>\n",
       "      <th>pc3</th>\n",
       "      <th>pc4</th>\n",
       "      <th>pc5</th>\n",
       "      <th>pc6</th>\n",
       "      <th>pc7</th>\n",
       "      <th>pc8</th>\n",
       "      <th>pc9</th>\n",
       "      <th>pc10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1001251</td>\n",
       "      <td>1001251</td>\n",
       "      <td>0</td>\n",
       "      <td>51</td>\n",
       "      <td>24.4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.452362</td>\n",
       "      <td>3.570649</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-14.2457</td>\n",
       "      <td>3.96257</td>\n",
       "      <td>-1.018640</td>\n",
       "      <td>2.277420</td>\n",
       "      <td>-4.21549</td>\n",
       "      <td>-0.100448</td>\n",
       "      <td>1.524450</td>\n",
       "      <td>-1.246880</td>\n",
       "      <td>1.368960</td>\n",
       "      <td>2.570260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1001445</td>\n",
       "      <td>1001445</td>\n",
       "      <td>1</td>\n",
       "      <td>63</td>\n",
       "      <td>30.2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.973000</td>\n",
       "      <td>2.930000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-13.4369</td>\n",
       "      <td>3.02070</td>\n",
       "      <td>-3.195770</td>\n",
       "      <td>0.428490</td>\n",
       "      <td>6.21226</td>\n",
       "      <td>-1.478700</td>\n",
       "      <td>1.282360</td>\n",
       "      <td>-1.606150</td>\n",
       "      <td>0.199501</td>\n",
       "      <td>-4.308460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1003040</td>\n",
       "      <td>1003040</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>31.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.291000</td>\n",
       "      <td>2.345000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-13.4688</td>\n",
       "      <td>4.63916</td>\n",
       "      <td>-2.339450</td>\n",
       "      <td>0.502712</td>\n",
       "      <td>-2.63828</td>\n",
       "      <td>0.662364</td>\n",
       "      <td>0.931805</td>\n",
       "      <td>0.987020</td>\n",
       "      <td>-5.069820</td>\n",
       "      <td>2.115390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1004201</td>\n",
       "      <td>1004201</td>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>36.8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.465000</td>\n",
       "      <td>4.716000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-11.7176</td>\n",
       "      <td>3.71645</td>\n",
       "      <td>-0.737745</td>\n",
       "      <td>1.572150</td>\n",
       "      <td>-1.20120</td>\n",
       "      <td>-1.681440</td>\n",
       "      <td>1.290020</td>\n",
       "      <td>-0.223860</td>\n",
       "      <td>-5.693960</td>\n",
       "      <td>-1.878020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1004219</td>\n",
       "      <td>1004219</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>37.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.483000</td>\n",
       "      <td>3.084000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-12.1573</td>\n",
       "      <td>2.01404</td>\n",
       "      <td>-2.064440</td>\n",
       "      <td>1.280780</td>\n",
       "      <td>-1.68080</td>\n",
       "      <td>2.864480</td>\n",
       "      <td>2.142890</td>\n",
       "      <td>-0.380478</td>\n",
       "      <td>-4.603690</td>\n",
       "      <td>0.058457</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       FID      IID  sex  age   bmi  ever_smoked  hdl_cholesterol  \\\n",
       "2  1001251  1001251    0   51  24.4          1.0         1.452362   \n",
       "3  1001445  1001445    1   63  30.2          1.0         0.973000   \n",
       "5  1003040  1003040    0   57  31.7          1.0         1.291000   \n",
       "8  1004201  1004201    0   55  36.8          1.0         1.465000   \n",
       "9  1004219  1004219    0   45  37.1          0.0         1.483000   \n",
       "\n",
       "   ldl_cholesterol  hypertension  ischemia_cc      pc1      pc2       pc3  \\\n",
       "2         3.570649             0            1 -14.2457  3.96257 -1.018640   \n",
       "3         2.930000             1            1 -13.4369  3.02070 -3.195770   \n",
       "5         2.345000             1            1 -13.4688  4.63916 -2.339450   \n",
       "8         4.716000             0            1 -11.7176  3.71645 -0.737745   \n",
       "9         3.084000             1            1 -12.1573  2.01404 -2.064440   \n",
       "\n",
       "        pc4      pc5       pc6       pc7       pc8       pc9      pc10  \n",
       "2  2.277420 -4.21549 -0.100448  1.524450 -1.246880  1.368960  2.570260  \n",
       "3  0.428490  6.21226 -1.478700  1.282360 -1.606150  0.199501 -4.308460  \n",
       "5  0.502712 -2.63828  0.662364  0.931805  0.987020 -5.069820  2.115390  \n",
       "8  1.572150 -1.20120 -1.681440  1.290020 -0.223860 -5.693960 -1.878020  \n",
       "9  1.280780 -1.68080  2.864480  2.142890 -0.380478 -4.603690  0.058457  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_phenotype.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select only samples that have imputed data available and save phenotype table as CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get imputed data\n",
    "path_to_impute_file = f'/mnt/project/Bulk/Imputation/{imputation_folder}/ukb{imputation_field_id}_c1_b0_v1.sample'\n",
    "sample_file = pd.read_csv(\n",
    "    path_to_impute_file,\n",
    "    delimiter='\\s',\n",
    "    header=0,\n",
    "    names=['FID', 'IID', 'missing', 'sex'],\n",
    "    engine='python',\n",
    ")\n",
    "# Intersect the phenotype file and the imputed .sample file\n",
    "# to generate phenotype DataFrame for only samples included in the imputed data\n",
    "ischemia_df = df_phenotype.join(\n",
    "    sample_file.set_index('IID'), on='IID', rsuffix='_sample', how='inner'\n",
    ")\n",
    "# Drop unuseful columns from .fam file\n",
    "ischemia_df.drop(\n",
    "    columns=['FID_sample', 'missing', 'sex_sample'],\n",
    "    axis=1,\n",
    "    inplace=True,\n",
    "    errors='ignore',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write phenotype files to a TSV file\n",
    "ischemia_df.to_csv('ischemia_df.phe', sep='\\t', na_rep='NA', index=False, quoting=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FID</th>\n",
       "      <th>IID</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>bmi</th>\n",
       "      <th>ever_smoked</th>\n",
       "      <th>hdl_cholesterol</th>\n",
       "      <th>ldl_cholesterol</th>\n",
       "      <th>hypertension</th>\n",
       "      <th>ischemia_cc</th>\n",
       "      <th>pc1</th>\n",
       "      <th>pc2</th>\n",
       "      <th>pc3</th>\n",
       "      <th>pc4</th>\n",
       "      <th>pc5</th>\n",
       "      <th>pc6</th>\n",
       "      <th>pc7</th>\n",
       "      <th>pc8</th>\n",
       "      <th>pc9</th>\n",
       "      <th>pc10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1001251</td>\n",
       "      <td>1001251</td>\n",
       "      <td>0</td>\n",
       "      <td>51</td>\n",
       "      <td>24.4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.452362</td>\n",
       "      <td>3.570649</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-14.2457</td>\n",
       "      <td>3.96257</td>\n",
       "      <td>-1.018640</td>\n",
       "      <td>2.277420</td>\n",
       "      <td>-4.21549</td>\n",
       "      <td>-0.100448</td>\n",
       "      <td>1.524450</td>\n",
       "      <td>-1.246880</td>\n",
       "      <td>1.368960</td>\n",
       "      <td>2.570260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1001445</td>\n",
       "      <td>1001445</td>\n",
       "      <td>1</td>\n",
       "      <td>63</td>\n",
       "      <td>30.2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.973000</td>\n",
       "      <td>2.930000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-13.4369</td>\n",
       "      <td>3.02070</td>\n",
       "      <td>-3.195770</td>\n",
       "      <td>0.428490</td>\n",
       "      <td>6.21226</td>\n",
       "      <td>-1.478700</td>\n",
       "      <td>1.282360</td>\n",
       "      <td>-1.606150</td>\n",
       "      <td>0.199501</td>\n",
       "      <td>-4.308460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1003040</td>\n",
       "      <td>1003040</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>31.7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.291000</td>\n",
       "      <td>2.345000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-13.4688</td>\n",
       "      <td>4.63916</td>\n",
       "      <td>-2.339450</td>\n",
       "      <td>0.502712</td>\n",
       "      <td>-2.63828</td>\n",
       "      <td>0.662364</td>\n",
       "      <td>0.931805</td>\n",
       "      <td>0.987020</td>\n",
       "      <td>-5.069820</td>\n",
       "      <td>2.115390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1004201</td>\n",
       "      <td>1004201</td>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>36.8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.465000</td>\n",
       "      <td>4.716000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-11.7176</td>\n",
       "      <td>3.71645</td>\n",
       "      <td>-0.737745</td>\n",
       "      <td>1.572150</td>\n",
       "      <td>-1.20120</td>\n",
       "      <td>-1.681440</td>\n",
       "      <td>1.290020</td>\n",
       "      <td>-0.223860</td>\n",
       "      <td>-5.693960</td>\n",
       "      <td>-1.878020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1004219</td>\n",
       "      <td>1004219</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>37.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.483000</td>\n",
       "      <td>3.084000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-12.1573</td>\n",
       "      <td>2.01404</td>\n",
       "      <td>-2.064440</td>\n",
       "      <td>1.280780</td>\n",
       "      <td>-1.68080</td>\n",
       "      <td>2.864480</td>\n",
       "      <td>2.142890</td>\n",
       "      <td>-0.380478</td>\n",
       "      <td>-4.603690</td>\n",
       "      <td>0.058457</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       FID      IID  sex  age   bmi  ever_smoked  hdl_cholesterol  \\\n",
       "2  1001251  1001251    0   51  24.4          1.0         1.452362   \n",
       "3  1001445  1001445    1   63  30.2          1.0         0.973000   \n",
       "5  1003040  1003040    0   57  31.7          1.0         1.291000   \n",
       "8  1004201  1004201    0   55  36.8          1.0         1.465000   \n",
       "9  1004219  1004219    0   45  37.1          0.0         1.483000   \n",
       "\n",
       "   ldl_cholesterol  hypertension  ischemia_cc      pc1      pc2       pc3  \\\n",
       "2         3.570649             0            1 -14.2457  3.96257 -1.018640   \n",
       "3         2.930000             1            1 -13.4369  3.02070 -3.195770   \n",
       "5         2.345000             1            1 -13.4688  4.63916 -2.339450   \n",
       "8         4.716000             0            1 -11.7176  3.71645 -0.737745   \n",
       "9         3.084000             1            1 -12.1573  2.01404 -2.064440   \n",
       "\n",
       "        pc4      pc5       pc6       pc7       pc8       pc9      pc10  \n",
       "2  2.277420 -4.21549 -0.100448  1.524450 -1.246880  1.368960  2.570260  \n",
       "3  0.428490  6.21226 -1.478700  1.282360 -1.606150  0.199501 -4.308460  \n",
       "5  0.502712 -2.63828  0.662364  0.931805  0.987020 -5.069820  2.115390  \n",
       "8  1.572150 -1.20120 -1.681440  1.290020 -0.223860 -5.693960 -1.878020  \n",
       "9  1.280780 -1.68080  2.864480  2.142890 -0.380478 -4.603690  0.058457  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ischemia_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load file to project storage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file-J0G5yy0JjKYBzvP4Vx26PbG0\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$output_dir\"\n",
    "# Upload the geno-pheno intersect phenotype file back to the RAP project\n",
    "dx upload ischemia_df.phe -p --path $1 --brief"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is an example of phenotype file.\n",
    "\n",
    "*Note: this table contains synthetic data and is made for demonstration purposes only!*\n",
    "\n",
    "|Row|FID|IID|sex|age|bmi|ever_smoked|hypertension|hdl_cholesterol|ldl_cholesterol|ischemia_cc|\n",
    "|:--- |:--- |:---|:--- |:---|:---|:---|:---|:--- |:---|:---|\n",
    "|0|EID-XXXXXXX|1234567|0|61.0|28.5|0|1|0.959|2.456|1|1|1|\n",
    "|1|EID-YYYYYYY|1234568|1|51.0|24.5|0|0|1.159|2.719|0|1|1|\n",
    "|2|EID-ZZZZZZZ|1234569|0|60.0|23.8|1|1|1.579|3.396|0|1|1|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output files\n",
    "\n",
    "- Table containing phenotype and covariates to be used in regenie GWAS analysis (`ischemia_df.phe`)"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,py"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
